File: macos/Sources/AFLEdgeMac/Models/APIModels.swift
import Foundation

struct HealthResponse: Codable, Equatable {
    var status: String
    var databaseOk: Bool
    var lastSuccessfulImportAt: Date?
}

struct DataFileStatus: Codable, Equatable, Identifiable {
    var fileName: String
    var relativePath: String
    var modifiedAt: Date

    var id: String { relativePath }
}

struct DataFileSection: Codable, Equatable, Identifiable {
    var code: String
    var title: String
    var category: String
    var files: [DataFileStatus]

    var id: String { code }
}

struct DataStatusResponse: Codable, Equatable {
    var generatedAt: Date
    var sections: [DataFileSection]
}

struct TeamSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var name: String
}

struct PlayerSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var fullName: String
}

struct PlayerStatOption: Codable, Equatable, Identifiable, Hashable {
    var code: String
    var label: String

    var id: String { code }
}

struct PlayerStatFilterOptions: Codable, Equatable, Identifiable {
    var playerId: Int
    var playerName: String
    var stats: [PlayerStatOption]
    var seasons: [String]
    var oppositions: [String]
    var venues: [String]
    var weatherCategories: [String]
    var homeAwayOptions: [String]

    var id: Int { playerId }
}

struct PlayerGameLogEntry: Codable, Equatable, Identifiable {
    var date: Date
    var roundLabel: String?
    var home: String?
    var venue: String?
    var weather: String?
    var away: String?
    var team: String?
    var opposition: String?
    var margin: Int?
    var tog: Double?
    var disposals: Double?
    var kicks: Double?
    var handballs: Double?
    var marks: Double?
    var goals: Double?
    var behinds: Double?
    var tackles: Double?
    var hitouts: Double?
    var freesFor: Double?
    var freesAgainst: Double?
    var fantasy: Double?
    var cba: Double?
    var gameNumber: Int
    var selectedStat: String
    var selectedValue: Double?
    var hit: Bool?

    var id: String { "\(gameNumber)-\(date.timeIntervalSince1970)-\(selectedStat)" }
}

struct PlayerStatSummary: Codable, Equatable {
    var playerId: Int
    var statCode: String
    var statLabel: String
    var lineMode: String
    var referenceLine: Double?
    var lowerBound: Double?
    var upperBound: Double?
    var sampleSize: Int
    var proportionOver: Double?
    var proportionUnder: Double?
    var impliedOddsOver: Double?
    var impliedOddsUnder: Double?
    var proportionWithinInterval: Double?
    var proportionOutsideInterval: Double?
    var impliedOddsWithinInterval: Double?
    var impliedOddsOutsideInterval: Double?
}

struct BookmakerSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var code: String
    var displayName: String
    var enabled: Bool
    var livePricingEnabled: Bool
    var sgmEligibleCount: Int
}

struct EventSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var matchName: String
    var startTime: Date?
    var roundLabel: String?
    var venue: String?
    var homeTeam: TeamSummary
    var awayTeam: TeamSummary
    var availableBookmakers: [String]
}

struct MarketSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var eventId: Int
    var marketTypeCode: String
    var displayName: String
    var player: PlayerSummary?
    var lineValue: Double?
    var bookmaker: String
    var availableSelectionTypes: [String]
}

struct SelectionSummary: Codable, Equatable, Identifiable, Hashable {
    var id: Int
    var marketId: Int
    var selectionType: String
    var label: String
    var decimalPrice: Double?
    var impliedProb: Double?
    var bookmaker: String
    var sgmEligible: Bool
    var edgePct: Double?
}

struct WeatherSummary: Codable, Equatable, Hashable {
    var temperatureC: Double?
    var windKph: Double?
    var precipProbability: Double?
    var precipMm: Double?
    var label: String?
    var iconCode: String?
}

struct OddsSearchResult: Codable, Equatable, Identifiable, Hashable {
    var selectionId: Int
    var marketId: Int
    var eventId: Int
    var matchName: String
    var startTime: Date?
    var bookmaker: String
    var marketTypeCode: String
    var marketDisplayName: String
    var player: PlayerSummary?
    var selectionType: String
    var label: String
    var lineValue: Double?
    var decimalPrice: Double?
    var impliedProb: Double?
    var edgePct: Double?
    var diff2025: Double?
    var diffLast10: Double?
    var playerPosition: String?
    var matchupDifficulty: String?
    var weather: WeatherSummary?
    var isBestPrice: Bool
    var nextBestProbDiff: Double?
    var sgmEligible: Bool

    var id: Int { selectionId }
}

struct PropSearchResult: Codable, Equatable, Identifiable, Hashable {
    var selectionId: Int
    var eventId: Int
    var matchName: String
    var startTime: Date?
    var bookmaker: String
    var marketTypeCode: String
    var player: PlayerSummary?
    var selectionType: String
    var label: String
    var lineValue: Double?
    var decimalPrice: Double?
    var impliedProb: Double?
    var edgePct: Double?
    var sgmEligible: Bool

    var id: Int { selectionId }
}

struct RequestedLeg: Codable, Equatable {
    var selectionId: Int
}

struct QuoteLeg: Codable, Equatable, Identifiable, Hashable {
    var selectionId: Int
    var label: String
    var marketTypeCode: String
    var selectionType: String
    var basePrice: Double

    var id: Int { selectionId }
}

struct SgmQuoteRequestPayload: Codable, Equatable {
    var bookmaker: String
    var eventId: Int
    var legs: [RequestedLeg]
    var forceRefresh: Bool
}

struct SgmQuoteResponse: Codable, Equatable, Identifiable {
    var quoteId: String
    var bookmaker: String
    var eventId: Int
    var legs: [QuoteLeg]
    var unadjustedPrice: Double
    var quotedPrice: Double
    var adjustmentFactor: Double
    var fromCache: Bool
    var quotedAt: Date
    var expiresAt: Date
    var status: String

    var id: String { quoteId }
}

struct SgmCompareRequestPayload: Codable, Equatable {
    var eventId: Int
    var selectionIds: [Int]
    var forceRefresh: Bool
}

struct SgmAgencyComparison: Codable, Equatable, Identifiable {
    var quoteId: String
    var bookmaker: String
    var eventId: Int
    var legs: [QuoteLeg]
    var unadjustedPrice: Double
    var quotedPrice: Double
    var adjustmentFactor: Double
    var fromCache: Bool
    var quotedAt: Date
    var expiresAt: Date
    var status: String

    var id: String { "\(bookmaker)-\(quoteId)" }
}

struct SgmCompareResponse: Codable, Equatable {
    var eventId: Int
    var selectionCount: Int
    var results: [SgmAgencyComparison]
}

struct CgmCompareRequestPayload: Codable, Equatable {
    var selectionIds: [Int]
}

struct CgmLegPrice: Codable, Equatable, Identifiable, Hashable {
    var selectionId: Int
    var matchName: String
    var label: String
    var marketTypeCode: String
    var selectionType: String
    var basePrice: Double

    var id: Int { selectionId }
}

struct CgmAgencyComparison: Codable, Equatable, Identifiable {
    var bookmaker: String
    var quotedPrice: Double
    var selectionCount: Int
    var legs: [CgmLegPrice]

    var id: String { bookmaker }
}

struct CgmCompareResponse: Codable, Equatable {
    var selectionCount: Int
    var results: [CgmAgencyComparison]
}

struct APIErrorEnvelope: Codable, Equatable {
    var error: APIErrorBody
}

struct APIErrorBody: Codable, Equatable {
    var code: String
    var message: String
    var retriable: Bool
    var details: [String: String]?
}

File: android/app/src/main/java/com/jamesbrown/aflmobile/model/ApiModels.kt
package com.jamesbrown.aflmobile.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement


@Serializable
data class HealthResponse(
    val status: String,
    @SerialName("database_ok") val databaseOk: Boolean,
    @SerialName("last_successful_import_at") val lastSuccessfulImportAt: String? = null,
)

@Serializable
data class DataFileStatus(
    @SerialName("file_name") val fileName: String,
    @SerialName("relative_path") val relativePath: String,
    @SerialName("modified_at") val modifiedAt: String,
)

@Serializable
data class DataFileSection(
    val code: String,
    val title: String,
    val category: String,
    val files: List<DataFileStatus>,
)

@Serializable
data class DataStatusResponse(
    @SerialName("generated_at") val generatedAt: String,
    val sections: List<DataFileSection>,
)

@Serializable
data class TeamSummary(
    val id: Int,
    val name: String,
)

@Serializable
data class PlayerSummary(
    val id: Int,
    @SerialName("full_name") val fullName: String,
)

@Serializable
data class PlayerStatOption(
    val code: String,
    val label: String,
)

@Serializable
data class PlayerStatFilterOptions(
    @SerialName("player_id") val playerId: Int,
    @SerialName("player_name") val playerName: String,
    val stats: List<PlayerStatOption>,
    val seasons: List<String>,
    val oppositions: List<String>,
    val venues: List<String>,
    @SerialName("weather_categories") val weatherCategories: List<String>,
    @SerialName("home_away_options") val homeAwayOptions: List<String>,
)

@Serializable
data class PlayerGameLogEntry(
    val date: String,
    @SerialName("round_label") val roundLabel: String? = null,
    val home: String? = null,
    val venue: String? = null,
    val weather: String? = null,
    val away: String? = null,
    val team: String? = null,
    val opposition: String? = null,
    val margin: Int? = null,
    val tog: Double? = null,
    val disposals: Double? = null,
    val kicks: Double? = null,
    val handballs: Double? = null,
    val marks: Double? = null,
    val goals: Double? = null,
    val behinds: Double? = null,
    val tackles: Double? = null,
    val hitouts: Double? = null,
    @SerialName("frees_for") val freesFor: Double? = null,
    @SerialName("frees_against") val freesAgainst: Double? = null,
    val fantasy: Double? = null,
    val cba: Double? = null,
    @SerialName("game_number") val gameNumber: Int,
    @SerialName("selected_stat") val selectedStat: String,
    @SerialName("selected_value") val selectedValue: Double? = null,
    val hit: Boolean? = null,
)

@Serializable
data class PlayerStatSummary(
    @SerialName("player_id") val playerId: Int,
    @SerialName("stat_code") val statCode: String,
    @SerialName("stat_label") val statLabel: String,
    @SerialName("line_mode") val lineMode: String,
    @SerialName("reference_line") val referenceLine: Double? = null,
    @SerialName("lower_bound") val lowerBound: Double? = null,
    @SerialName("upper_bound") val upperBound: Double? = null,
    @SerialName("sample_size") val sampleSize: Int,
    @SerialName("proportion_over") val proportionOver: Double? = null,
    @SerialName("proportion_under") val proportionUnder: Double? = null,
    @SerialName("implied_odds_over") val impliedOddsOver: Double? = null,
    @SerialName("implied_odds_under") val impliedOddsUnder: Double? = null,
    @SerialName("proportion_within_interval") val proportionWithinInterval: Double? = null,
    @SerialName("proportion_outside_interval") val proportionOutsideInterval: Double? = null,
    @SerialName("implied_odds_within_interval") val impliedOddsWithinInterval: Double? = null,
    @SerialName("implied_odds_outside_interval") val impliedOddsOutsideInterval: Double? = null,
)

@Serializable
data class BookmakerSummary(
    val id: Int,
    val code: String,
    @SerialName("display_name") val displayName: String,
    val enabled: Boolean,
    @SerialName("live_pricing_enabled") val livePricingEnabled: Boolean,
    @SerialName("sgm_eligible_count") val sgmEligibleCount: Int,
)

@Serializable
data class EventSummary(
    val id: Int,
    @SerialName("match_name") val matchName: String,
    @SerialName("start_time") val startTime: String? = null,
    @SerialName("round_label") val roundLabel: String? = null,
    val venue: String? = null,
    @SerialName("home_team") val homeTeam: TeamSummary,
    @SerialName("away_team") val awayTeam: TeamSummary,
    @SerialName("available_bookmakers") val availableBookmakers: List<String>,
)

@Serializable
data class MarketSummary(
    val id: Int,
    @SerialName("event_id") val eventId: Int,
    @SerialName("market_type_code") val marketTypeCode: String,
    @SerialName("display_name") val displayName: String,
    val player: PlayerSummary? = null,
    @SerialName("line_value") val lineValue: Double? = null,
    val bookmaker: String,
    @SerialName("available_selection_types") val availableSelectionTypes: List<String>,
)

@Serializable
data class SelectionSummary(
    val id: Int,
    @SerialName("market_id") val marketId: Int,
    @SerialName("selection_type") val selectionType: String,
    val label: String,
    @SerialName("decimal_price") val decimalPrice: Double? = null,
    @SerialName("implied_prob") val impliedProb: Double? = null,
    val bookmaker: String,
    @SerialName("sgm_eligible") val sgmEligible: Boolean,
    @SerialName("edge_pct") val edgePct: Double? = null,
)

@Serializable
data class WeatherSummary(
    @SerialName("temperature_c") val temperatureC: Double? = null,
    @SerialName("wind_kph") val windKph: Double? = null,
    @SerialName("precip_probability") val precipProbability: Double? = null,
    @SerialName("precip_mm") val precipMm: Double? = null,
    val label: String? = null,
    @SerialName("icon_code") val iconCode: String? = null,
)

@Serializable
data class OddsSearchResult(
    @SerialName("selection_id") val selectionId: Int,
    @SerialName("market_id") val marketId: Int,
    @SerialName("event_id") val eventId: Int,
    @SerialName("match_name") val matchName: String,
    @SerialName("start_time") val startTime: String? = null,
    val bookmaker: String,
    @SerialName("market_type_code") val marketTypeCode: String,
    @SerialName("market_display_name") val marketDisplayName: String,
    val player: PlayerSummary? = null,
    @SerialName("selection_type") val selectionType: String,
    val label: String,
    @SerialName("line_value") val lineValue: Double? = null,
    @SerialName("decimal_price") val decimalPrice: Double? = null,
    @SerialName("implied_prob") val impliedProb: Double? = null,
    @SerialName("edge_pct") val edgePct: Double? = null,
    @SerialName("diff_2025") val diff2025: Double? = null,
    @SerialName("diff_last_10") val diffLast10: Double? = null,
    @SerialName("player_position") val playerPosition: String? = null,
    @SerialName("matchup_difficulty") val matchupDifficulty: String? = null,
    val weather: WeatherSummary? = null,
    @SerialName("is_best_price") val isBestPrice: Boolean = false,
    @SerialName("next_best_prob_diff") val nextBestProbDiff: Double? = null,
    @SerialName("sgm_eligible") val sgmEligible: Boolean,
)

@Serializable
data class PropSearchResult(
    @SerialName("selection_id") val selectionId: Int,
    @SerialName("event_id") val eventId: Int,
    @SerialName("match_name") val matchName: String,
    @SerialName("start_time") val startTime: String? = null,
    val bookmaker: String,
    @SerialName("market_type_code") val marketTypeCode: String,
    val player: PlayerSummary? = null,
    @SerialName("selection_type") val selectionType: String,
    val label: String,
    @SerialName("line_value") val lineValue: Double? = null,
    @SerialName("decimal_price") val decimalPrice: Double? = null,
    @SerialName("implied_prob") val impliedProb: Double? = null,
    @SerialName("edge_pct") val edgePct: Double? = null,
    @SerialName("sgm_eligible") val sgmEligible: Boolean,
)

@Serializable
data class QuoteLeg(
    @SerialName("selection_id") val selectionId: Int,
    val label: String,
    @SerialName("market_type_code") val marketTypeCode: String,
    @SerialName("selection_type") val selectionType: String,
    @SerialName("base_price") val basePrice: Double,
)

@Serializable
data class SgmCompareRequestPayload(
    @SerialName("event_id") val eventId: Int,
    @SerialName("selection_ids") val selectionIds: List<Int>,
    @SerialName("force_refresh") val forceRefresh: Boolean,
)

@Serializable
data class SgmAgencyComparison(
    @SerialName("quote_id") val quoteId: String,
    val bookmaker: String,
    @SerialName("event_id") val eventId: Int,
    val legs: List<QuoteLeg>,
    @SerialName("unadjusted_price") val unadjustedPrice: Double,
    @SerialName("quoted_price") val quotedPrice: Double,
    @SerialName("adjustment_factor") val adjustmentFactor: Double,
    @SerialName("from_cache") val fromCache: Boolean,
    @SerialName("quoted_at") val quotedAt: String,
    @SerialName("expires_at") val expiresAt: String,
    val status: String,
)

@Serializable
data class SgmCompareResponse(
    @SerialName("event_id") val eventId: Int,
    @SerialName("selection_count") val selectionCount: Int,
    val results: List<SgmAgencyComparison>,
)

@Serializable
data class CgmCompareRequestPayload(
    @SerialName("selection_ids") val selectionIds: List<Int>,
)

@Serializable
data class CgmLegPrice(
    @SerialName("selection_id") val selectionId: Int,
    @SerialName("match_name") val matchName: String,
    val label: String,
    @SerialName("market_type_code") val marketTypeCode: String,
    @SerialName("selection_type") val selectionType: String,
    @SerialName("base_price") val basePrice: Double,
)

@Serializable
data class CgmAgencyComparison(
    val bookmaker: String,
    @SerialName("quoted_price") val quotedPrice: Double,
    @SerialName("selection_count") val selectionCount: Int,
    val legs: List<CgmLegPrice>,
)

@Serializable
data class CgmCompareResponse(
    @SerialName("selection_count") val selectionCount: Int,
    val results: List<CgmAgencyComparison>,
)

@Serializable
data class ApiErrorEnvelope(
    val error: ApiErrorBody,
)

@Serializable
data class ApiErrorBody(
    val code: String,
    val message: String,
    val retriable: Boolean,
    val details: JsonElement? = null,
)
